US7: Changes on the front-end & authentication

paulinezhanghh opened the pull request:
## Overview
This pull request implements **User Story 7**: “As a user, I want to see everyone’s full names so I know who I’m talking to on NodeBB as a username does not give much information about who this person is.” We added the full name text entry box on the user registration page and added restrictions for full names so we made sure that all the full names registered are valid.
## Features Implemented

### 1. Frontend Changes
- **Added Fullname Entry Box on Registration page**: Added an input box for fullname which allows users to put their fullnames in. This ensures that all the users will have a fullname.

### 2. Backend Changes
- Authentication During Registration: We added several restrictions while users are entering their fullname. For example, users are not allowed to enter leave it blank, or put all spaces in the input box. Also, they are not allowed to put URLs or to exceed 255 characters. 
- Authentication For Updating Profiles: To make sure that a user will always have a valid fullname, we also updated the authentication checks for the logic of updating the profile as it allowed the users to leave the fullname as blank initially.
- Files Changed For Authentication
   (1) **During Registration**: Updated authentication method in src/controllers/authentication.js by adding restrictions like no blank fullname.
   (2) **Edit Profile Page**: Similar to the first one, updated authentication method in src/user/profile.js.
- Error Messages: We added error messages like "Invalid Fullname" and "Fullname is Required" into the json file so that NodeBB can retrieve the correct error messages.
- Files Changed For Error Message
   (1) **Different Language Version**: In order to cover all the English-based environments, we updated the error message in the error.json file under 3 different language versions (en-GB, en-US, and en-x-pirate).

## Test Coverage
- **Comprehensive Tests**: We wrote tests to validate the authentication checks of fullname and make sure that the system can successfully store the fullname inputted to the original fullname attribute.
- **Edge Cases**: Tests were also added to cover edge cases, such as fullname with all blank spaces, fullname with a length of 256, and fullname as URL.

Screenshot:
<img width="437" alt="Screenshot 2024-09-21 at 19 14 36" src="https://github.com/user-attachments/assets/7ebacea7-c2ce-4399-916e-41258cd02800">


Solves #30 #31 #32

e5z26 committed:
Update attributes for fullname

Added 2 attributes for creating full-name place on register page - "fullname" and "fullname-placeholder" are the text that will be added around the input box for fullname

paulinezhanghh committed:
Make register.tpl tracked file

Use git add -f to make the register.tpl a tracked file so we can edit it

paulinezhanghh committed:
Front-end changes & authentication

We added the text entry box and the authentication to make sure that all the users will enter their full name while registering.

paulinezhanghh committed:
Merge branch 'f24' into us-7

e5z26 committed:
Added space between function and brackets

Added space between function and brackets in authentication.js

e5z26 committed:
Merge branch 'us-7' of https://github.com/CMU-313/nodebb-f24-team-five-guys into us-7

paulinezhanghh committed:
Update authentication.js

paulinezhanghh committed:
Merge branch 'us-7' of https://github.com/CMU-313/nodebb-f24-team-five-guys into us-7

paulinezhanghh committed:
Revise tests about user registration with fullnames

Added fullname attribute for a lot of test cases in 3 files: authentication.js, controller.js, and user.js to pass the tests.

paulinezhanghh committed:
Revised email-related tests

Since email-related tests are also affected by adding a full name, we added fullname attribute for a lot of the cases in emails.js to pass the test.

paulinezhanghh committed:
Minor changes for lint test

Fixed some errors for the lint test.

paulinezhanghh committed:
Add new checks for US7

There are some additional checks we want to add for US7. For example, there should be a fullname-required error pops up if fullname is not properly entered.

LeeMEow23 reviewed (approved):
The newly added code looks well-structured! Thanks for the great work, approved!

paulinezhanghh committed:
revise the test for us7

Change the description of the test

e5z26 committed:
Updated errors and restrictions

In public/language/en-US/error.json, added displayed text for error fullname-required

In src/controllers/authentication.js, added more checks on what the full name cannot be (to match up with the existing checks when setting up full name in profile), also specified which error should be called.

In src/user/profile.js, updated existing checks on fullname to match with full name set during registration, also ensured that there is a full name

paulinezhanghh committed:
Add more Checks for fullname test

Added more checks for fullname. For example, when user enters a URL as their fullname, there should be an invalid fullname error pops up. Also, revised the profile creation file accordingly.

paulinezhanghh committed:
Adding restrictions for updating fullname in profile

To keep the current profile updating logic the same as the fullname entered for registration, the user are not allowed to leave the fullname blank when updating the fullname in their profile.

paulinezhanghh committed:
changes for lint test

add a space before the opening brace

paulinezhanghh committed:
deleting something to test

There is one unchanged line that lost coverage; we are trying to find which line it is.

paulinezhanghh committed:
Revert "deleting something to test"

This reverts commit 0b6eb5d065aa04d99493814661d54176601045c3.

paulinezhanghh committed:
Get back to the working version

Due to some reasons, there is one unchanged line that lost coverage; we want to delete the changes we made to this file and see if everything works fine.

e5z26 committed:
Added condition for needing a full name for update profile

src/user/profile.js is updated to check if newfullname is valid

also added error messages

e5z26 committed:
added test cases for updating full name in profile

added test cases in src/user/profile.js to check for when updating full name

paulinezhanghh committed:
delete unnecessary check

We added this check last week but realized it should not be checked as we didn't add anything about this.

paulinezhanghh committed:
adding comments to checks

added new check added comments to all the checks

joeni21 reviewed (approved):
The code looks clean and the checks cover everything. Not a big issue but maybe remove the unnecessary comment "new check added". Might be a bit confusing in future, but overall looks great.
